### ISSUE TITLE
WeightedRandomSampler with replacement=False incredibly slow

### ISSUE THREAD

**Issue by michielkallenberg** [pytorch/pytorch#32221]
## 🐛 Bug

WeightedRandomSampler with replacement=False incredibly slow

## To Reproduce

``
import torch,torchvision
list(torch.utils.data.WeightedRandomSampler([0.1, 0.9, 0.4, 0.7, 3.0, 0.6]*10000000, 500, replacement=False))
``
## Expected behavior

Above code takes very long to run, whereas with replacement=True this is more or less instantaneous. Is this a known issue, and what can be done to overcome this?

## Environment
 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

## Additional context

cc @VitalyFedyunin @ngimel @mruberry

**Comment by mruberry** [comment]
Thanks for reporting this issue! I'm not sure if it was previously known, although it is not especially surprising since replacement=False is a stronger requirement than replacement=True, and it could be challenging to parallelize. If you're interested in improving the replacement=False algorithm it would be interesting to review a PR!

If you're depending on replacement=False to be faster, then maybe you could workaround the issue by generating a mask from a distribution?

**Comment by zimenglan-sysu-512** [comment]
@michielkallenberg 
have u solve the problem? if the size of weights is to larger, and set 'replacement' False, will cost too much time.

**Comment by lsawade** [comment]
I just tried this on a macOS, and it's near-instantaneous for both options [`True`/`False`]. I tested it because I was running the same line of code on a PowerPC cluster, on which this also took forever (correction, still is taking forever... hasn't finished yet, and probably won't...).

It's important to note that the torch versions are not the same:

`macOS`: `1.10.0`
`Linux (ppc)`: `1.3.1` (Hard to upgrade this one, but I guess I'll have to try installing from source...)

**Comment by flpgrz** [comment]
It seems it's slow also when `replacement=False`.
Apparently, the time required for sampling scales with the size of the dataset.
I'm trying to balance 1899075 samples, of which 1747117 are negative, and 151958 are positive.

**Comment by ngimel** [comment]
Yes, this was fixed after 1.3
